src/serialization.cpp

```diff
@@ -197,11 +198,98 @@ void decompressZlib(std::istream &is, std::ostream &os, size_t limit)
 	inflateEnd(&z);
 }
 
-void compress(const SharedBuffer<u8> &data, std::ostream &os, u8 version)
+struct ZSTD_Deleter {
+	void operator() (ZSTD_CCtx* cctx) {
+		ZSTD_freeCCtx(cctx);
+	}
+
+	void operator() (ZSTD_DCtx* dctx) {
+		ZSTD_freeDCtx(dctx);
+	}
+};
+
+void compressZstd(const u8 *data, size_t data_size, std::ostream &os, int level)
 {
+	// reusing the context is recommended for performance
+	// it will destroyed when the thread ends
+	thread_local std::unique_ptr<ZSTD_CCtx, ZSTD_Deleter> cctx(ZSTD_createCCtx());
+
+	const size_t bufsize = ZSTD_CStreamOutSize();
+	char output_buffer[bufsize];
+
+	ZSTD_CCtx_setParameter(cctx.get(), ZSTD_c_compressionLevel, level);
+
+	ZSTD_inBuffer input = { data, data_size, 0 };
+	ZSTD_outBuffer output = { output_buffer, bufsize, 0 };
+	size_t ret;
+	do
+	{
+		ret = ZSTD_compressStream2(cctx.get(), &output , &input, ZSTD_e_end);
+		if (ZSTD_isError(ret)) {
+			dstream << ZSTD_getErrorName(ret) << std::endl;
+			throw SerializationError("compressZstd: failed");
+		}
+		os.write(output_buffer, output.pos);
+		output.pos = 0;
+	} while (ret != 0);
+}
+
+void compressZstd(const std::string &data, std::ostream &os, int level)
+{
+	compressZstd((u8*)data.c_str(), data.size(), os, level);
+}
+
+void decompressZstd(std::istream &is, std::ostream &os)
+{
+	// reusing the context is recommended for performance
+	// it will destroyed when the thread ends
+	thread_local std::unique_ptr<ZSTD_DCtx, ZSTD_Deleter> dctx(ZSTD_createDCtx());
+
+	const size_t bufsize_out = ZSTD_CStreamOutSize();
+	const size_t bufsize_in = ZSTD_CStreamInSize();
+	char output_buffer[bufsize_out];
+	char input_buffer[bufsize_in];
+
+	ZSTD_outBuffer output = { output_buffer, bufsize_out, 0 };
+	ZSTD_inBuffer input = { input_buffer, 0, 0 };
+	size_t ret;
+	do
+	{
+		is.read(input_buffer, bufsize_in);
+	        input.size = is.gcount();
+		input.pos = 0;
+
+		ret = ZSTD_decompressStream(dctx.get(), &output, &input);
+		if (ZSTD_isError(ret)) {
+			dstream << ZSTD_getErrorName(ret) << std::endl;
+			throw SerializationError("decompressZstd: failed");
+		}
+		os.write(output_buffer, output.pos);
+		output.pos = 0;
+	} while(ret != 0);
+
+	// Unget all the data that inflate didn't take
+	is.clear(); // Just in case EOF is set
+	for (u32 i = 0; i < input.size - input.pos; i++) {
+		is.unget();
+		if (is.fail() || is.bad())
+			throw SerializationError("decompressZlib: unget failed");
+	}
+}
+
+void compress(const SharedBuffer<u8> &data, std::ostream &os, u8 version, int level)
+{
+	if(version >= 29)
+	{
+		// map the zlib levels [-1,9]
+		// the range 1-10 is good enough, 0 is the default (currently 3)
```

Comment:
You may find this questionable.
Perhaps, since compress can now call multiple schemes, we ought to come with a general compression scale, perhaps -1 to 100, and then map that whatever range the specific algorithms support.

Or remove the config option altogether and use the defaults....

Reply:
Yet another option is to always use -1 for zlib, and use the zstd spectrum for the config instead.
zstd has a wide range compressions from very fast to very slow, and perhaps we'll never need a new one again.